### DISCUSSION TITLE
Summary of the issues about custom iptables rules

### DISCUSSION THREAD

**Discussion by akerouanton** [moby/moby#45524] [Ideas]
I’d like to summarize here all the issues related to the way Docker publishes ports on the host, and the security concerns it poses. Some of these issues are specifically about UFW, the others are more generic.

We’re aware these issues have been a real pain for many, hence we want to address them in the near future. We’re going to undertake actions to both address the root cause and improve the documentation. Here’s a summary of the next sections:

- Current state
    1. The way Docker published ports
    2. The way `--publish` is used
    3. The `DOCKER-USER` chain and why it’s impracticable
    4. Existing solutions from the community
- Path forward
    1. A new port mapper
    2. A new hook point for custom port mappers
    3. Daemon-wide default value for `com.docker.network.bridge.host_binding_ipv4`
    4. Docs improvement

For the record, here’s the list of GH issues related to this matter:

- https://github.com/moby/moby/issues/22054
- https://github.com/moby/moby/issues/4737
- https://github.com/moby/moby/issues/42563
- https://github.com/moby/moby/issues/29312
- https://github.com/moby/moby/issues/1855
- https://github.com/moby/moby/issues/1251
- https://github.com/moby/moby/issues/27913
- https://github.com/docker/for-linux/issues/777
- https://github.com/docker/for-linux/issues/690

There's also been an issue filed on ufw's bug tracker about this problem:

- https://bugs.launchpad.net/ufw/+bug/1860677

I’d like to invite everyone with insights, advice or anything else to chime in.

## Current state

Currently, users who want to publish container ports to their host or to the outside world have to use `docker run --publish` (and it’s Compose equivalent). Most of the time users don’t specify what host IP address their port should be bound to (eg. `--publish 127.0.0.1:3306:3306`). The way Docker implement this feature, and the way users use it make it an unsafe feature.

### 1. The way Docker publishes ports

To publish ports, Docker uses `iptables` NAT rules to forward traffic coming to the host to the right container.

`iptables` has two main features: traffic NATing and packet filtering. Its design make it so that using its first main feature influence the way the second one works, and both features are intermingled. Said another way, when iptables apply a NAT rule to a packet, it diverts this packet from its `filter input` / `filter output` chains.

Many Docker users expect rules in `filter input` / `filter output` chains to be evaluated for all incoming traffic, *including* traffic going to containers. Unfortunately, that’s not how things work.

As a consequence, when users set their own custom iptables rules, either by using `iptables` directly or when they rely on an iptables frontend like ufw, Docker breaks their expectation and they end up being unprotected.

A few workarounds have been devised over the years to overcome this issue, but: 1. it’s mostly undocumented on Docker side (for various reasons) ; 2. some are quite contrived and one can get them wrong easily (as with anything involving iptables anyway).

### 2. The way `--publish` is used

The other part of the issue is really about how `--publish` is used. Most tutorials, blog posts, etc… and even Docker’s documentation doesn’t clearly state that `--publish 3306:3306` is ***insecure* *in most cases**.* That is, when someone use the above `--publish` flag, in most cases they ask Docker to actually open the port to the outside world.

When users want to publish a port only to access it from the host, they should use `--publish 127.0.0.1:3306:3306` instead.

Note that, “in most cases” refers to the daemon’s `--ip` flag (equivalent to the `ip` field in `daemon.json`), and the bridge option `com.docker.network.bridge.host_binding_ipv4`. The daemon flag set the default IP address published ports are bound to for containers attached to the default bridge (ie. `docker0`). The bridge option achieves the same, but for custom bridge networks. For instance:

```bash
$ docker network create -o com.docker.network.bridge.host_binding_ipv4=127.0.0.1 testnet
$ docker run --rm -d --network testnet -p 8080:80 traefik/whoami
$ ss -tlpn | grep 8080
LISTEN 0      4096            127.0.0.1:8080       0.0.0.0:*

# Following command is equivalent to the default behavior
$ docker run --rm -d --network testnet -p 0.0.0.0:80:80 traefik/whoami
$ ss -tlpn | grep :80  
LISTEN 0      4096              0.0.0.0:80         0.0.0.0:*
```

There’s currently no way to set a default value at the daemon level for all networks. One have to *always* specify it when creating new networks (more on that later).

*Note that we’re aware people have reported security issues related to `--publish 127.0.0.1:8080:80`. This is something we’re going to address in the near future too. However, this issue is not the right place to discuss those concerns.*

### 3. The `DOCKER-USER` chain and why it’s impracticable

Docker v17.06 introduced the `DOCKER-USER` chain. It’s meant to help users create their own iptables rules and make them evaluated before Docker’s own rules (as [[documented here](https://docs.docker.com/network/iptables/)](https://docs.docker.com/network/iptables/)). However there’re three major pitfalls.

First, this chain is part of the `forward` table. The direct consequence is that, at this stage, packets have already been DNATed. That is, users can only match containers’ internal IP and port with the usual iptables flags. To circumvent that, users have to either use static IP address allocation or run some daemon listening for Docker events (through [the API `/events` endpoint](https://docs.docker.com/engine/api/v1.42/#tag/System/operation/SystemEvents)) to dynamically create the appropriate iptables rules. This is unusable in practice.

If a user want to match traffic based on the original IP and port (as seen by their network interface), they have to leverage the `conntrack` iptables extension. And this is not without performance impact. For instance (see [the entry about conntrack extension in `man 8 iptables-extensions`](https://ipset.netfilter.org/iptables-extensions.man.html#lbAO)):

```bash
$ sudo iptables -I DOCKER-USER -p tcp -m conntrack --ctstate INVALID -j REJECT
$ sudo iptables -I DOCKER-USER 2 -p tcp -m conntrack --ctorigdst 192.168.121.249 --ctorigdstport 80 -j REJECT
```

The second pitfall, demonstrated by the above example, is that this chain actually works as an denylist by default, which is generally considered a bad practice when it comes to configuring firewalls. To turn it into an allowlist, one has to replace the `RETURN` rule added by Docker by a `DROP` / `REJECT` rule and then create the appropriate `RETURN` rules. Then, connections that don’t match user-defined rules will be dropped. For instance:

```bash
$ sudo iptables -t filter -F DOCKER-USER
$ sudo iptables -t filter -I DOCKER-USER -p tcp -m conntrack --ctstate ESTABLISHED -j RETURN
$ sudo iptables -t filter -I DOCKER-USER 2 -p tcp -m conntrack --ctorigdst 10.8.4.2 --ctorigdstport 80 -j RETURN
$ sudo iptables -t filter -A DOCKER-USER -j REJECT
```

The third pitfall with this chain is that: 1. it forces iptables frontend to provide a rule syntax to leverage the `conntrack` extension ; 2. it forces us to implement a way to redirect traffic to frontends’ own chains on demand (and frontends would have to do the same, to cover all edge-cases). This would be impracticable for maintainers, and still doesn’t cover well the case of user-defined custom iptables rules.

## Path forward

Now, how can we resolve this issue?

### 1. A new port mapper

Linux actually provides another way to implement DNATing: IPVS. With this NATing technique, packets would reach `filter input` / `filter output` chains and thus we’d not break user expectations anymore. As such, we need to implement a new IPVS-based portmapper.

However, it’s already been stated by maintainers that we can’t safely change the default portmapper without (probably) introducing BC breaks. As we don’t want that, we will:

1. Release this new portmapper as an experimental feature (gated by a feature flag), in a first time ; 
2. After a few releases, make it GA but still an opt-in. Keep the old iptables portmapper in the mean time.

Note that, although currently not publicly documented, moby/moby is leaning toward an accelerated release cadence. Thus, “After a few releases” is a matter of months, not years.

### 2. A new hook point for custom port mappers

The IPVS portmapper might not fit everyone’s needs. As we’re going to define a proper port mapper interface, we can also make it possible for users to run their own port mappers.

This would allow users to hook their shell script or binaries into docker. That would allow anyone to set their own way iptables rules, reconfigure their L4/L7 reverse proxy on the fly (eg. HAProxy), reconfigure some external load-balancer (eg. ALB/ELB), and what not.

### 3. Daemon-wide default value for `com.docker.network.bridge.host_binding_ipv4`

We might also need a global (ie. daemon-wide) config parameter to set the default value of the `com.docker.network.bridge.host_binding_ipv4` option for all networks. We can probably change the default value of this new config parameter when the IPVS portmapper is used. It would then default to `127.0.0.1` and thus users would have to make a conscious choice when publishing ports to the outside world.

### 4. Docs improvement

- Properly document the `host_binding_ipv4` option in [https://docs.docker.com/config/containers/container-networking/](https://docs.docker.com/config/containers/container-networking/)
- Also, make it loud and clear that `--publish` is insecure *by default* (same doc page, but probably others too).
- Clearly state the pitfalls of the `DOCKER-USER` chain in [https://docs.docker.com/network/iptables/](https://docs.docker.com/network/iptables/)
- And also document how to match packets based on their original IP/port destination (same doc page).
- Clearly state what steps people have to do to harden their networking security (ie. set `host_binding_ipv4` ; where?)
- ⬆️ Make that easy to find out in the installation guide. Sysops should jump to that page right after installing the daemon.
- Double-check other docs page, blog post, etc… to make sure we don’t advertise for an insecure feature (`—publish` without IP address prepended).
- Clearly state in Debian/Ubuntu installation guides that ufw and docker make conflicting use of iptables.

**Comment by patrakov** [comment]
Thanks for a well-written summary. The only obvious criticism from me is that it only lists issues filed against Docker, while I would expect people who thought that it is not a Docker fault to have filed issues against their favorite firewall frontend. Such issues (i.e. for ufw and firewalld) should also be linked from here.

**Reply by akerouanton** [reply]
> it only lists issues filed against Docker [...] Such issues (i.e. for ufw and firewalld) should also be linked from here.

I added a reference to the sole issue filed in ufw's bug tracker. For firewalld, dockerd should integrate with it well. If that's not the case, please file an issue so we can investigate that bug.

**Comment by taladar** [comment]
Whatever you do going forward, please stop acting as if Docker owns the system it runs on (as e.g. shown in the iptables based port mapping implementation it has now). Docker is one of many things running on the system and needs to integrate with other parts cleanly.

Docker needs functionality that is driven by the rest of the system, e.g. some command you can call after restoring firewall rules from elsewhere to restore Docker created rules or even better to allow hooking into the rule creation process so the admin has a way to filter and/or modify rules created by docker to fit their local needs.

> iptables has two main features: traffic NATing and packet filtering.

It has many other features as well, e.g. logging packets, rating limiting them, sending them to userspace for processing, fixing various parameters such as Path MTU/MSS, marking packets for policy routing and connection tracking,... Have a look at the iptables-extension man page some time. 

The point is that Docker can never hope to be a layer on top of iptables that offers everything the admin might need from iptables so it is best not to even try that.

Oh, and that doesn't even consider nftables yet which is becoming more and more relevant.

How about some sort of mechanism for a set of hook scripts to create/delete port forwarding which can have a simple default version and the admin can replace it with something more complicated as needed?

**Reply by akerouanton** [reply]
> How about some sort of mechanism for a set of hook scripts to create/delete port forwarding which can have a simple default version and the admin can replace it with something more complicated as needed?

Libnetwork has already not so great performance; shelling out to some external scripts by default would make this even worse. However, the hook points I'm proposing would meet this request.

> The point is that Docker can never hope to be a layer on top of iptables that offers everything the admin might need from iptables so it is best not to even try that.

We can't remove the current implementation without breaking many installations. We're inherently stuck with it for the time being. Hence, the proposal to: 1. add a new port mapper that don't mess as much with iptables ; 2. add a mechanism to let users write their own port mappers. Whether we make this new port mapper the default one is something we might consider in the future.

> Whatever you do going forward, please stop acting as if Docker owns the system it runs on (as e.g. shown in the iptables based port mapping implementation it has now). Docker is one of many things running on the system and needs to integrate with other parts cleanly.

FWIW, the IPVS port mapper will have a config parameter to let users specify the fwmarks ranges if the default one conflicts with something else.

> Docker needs functionality that is driven by the rest of the system, e.g. some command you can call after restoring firewall rules from elsewhere to restore Docker created rules

About this specific example, this is part of my todo-list. We need to overhaul the way iptables rules are created to add support for nftables. We'll consider adding a way to recreate dockerd's rules at that moment.

**Comment by BenjamenMeyer** [comment]
@akerouanton great summary and thank you helping drive this conversation.
Couple changes:

- [ ] You have a section on how `--publish` is used; it would be interesting to have a sub-section on how it was intended to be used; this would give understanding between intention and reality. The goal here is to help bridge the conversation between design and use, and help avoid similar things going forward.
- [ ] In section 2 on use of `--publish` it would be good to cover the many ways this is done between (a) [`docker run --publish`](https://docs.docker.com/engine/reference/commandline/run/#publish), (b) [Dockerfile `EXPOSE`](https://docs.docker.com/engine/reference/builder/#expose), and (c) [Docker Compose Application `ports`](https://docs.docker.com/compose/compose-file/05-services/#ports) as they all have the same effect.
- [ ] In section 3, please note how difficult it is for lay people to write IPTables rules correctly; this is kind of hinted at at the end of section 1.

**Comment by BenjamenMeyer** [comment]
Please consider how Kubernetes solved this issue too - using an Application Load Balancer to interface into the internal container networks. Traffic gets directed to the ALB to cross from host to containers, and then rules applied there for going to the containers themselves. This is cleaner than trying to deal with iptables/nftables/IPVS directly.

**Reply by BenjamenMeyer** [reply]
@akerouanton I'm pretty sure this issue is not solely a Linux issue though based on my analysis for #22054. It's been a while, but it would be good to have confirmation from others regarding Windows and Mac. So it would probably  be good to find a solution like an ALB approach that is more easily a cross-platform solution.

**Reply by corhere** [reply]
Are you referring to [`LoadBalancer` services?](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer)
> To implement a Service of `type: LoadBalancer`, Kubernetes typically starts off by making the changes that are equivalent to you requesting a Service of `type: NodePort`. The cloud-controller-manager component then configures the external load balancer to forward traffic to that assigned node port.

NodePort services are implemented by [`kube-proxy`](https://kubernetes.io/docs/concepts/overview/components/#kube-proxy).
> [kube-proxy](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) maintains network rules on nodes. These network rules allow network communication to your Pods from network sessions inside or outside of your cluster.
>
> **kube-proxy uses the operating system packet filtering layer** if there is one and it's available. Otherwise, kube-proxy forwards the traffic itself.

(Emphasis mine.)

> ```
> --proxy-mode ProxyMode
> ```
>
> [Which proxy mode to use: on Linux this can be 'iptables' (default) or 'ipvs'.](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/#:~:text=%2D%2Dproxy%2Dmode%20ProxyMode,default%29%20or%20%27ipvs%27.)

Application Load Balancers only handle routing of external traffic to a node. Kubernetes still depends on iptables/nftables/IPVS directly for crossing from host to containers. There is no performant cross-platform solution which Kubernetes is using that we've ignored.

**Comment by iTrooz** [comment]
This has been a year. I only learned about this pitfall now (and found out I have 2 databases exposed, thankfully with strong passwords), so I think docker should take more measures to teach their users about it. For example by merging https://github.com/moby/moby/pull/45076

**Comment by pikeas** [comment]
> We’re aware these issues have been a real pain for many, hence we want to address them in the near future.

@akerouanton This was posted ~16 months ago, can you share a progress update?

**Comment by melutovich** [comment]
Any update on this?
